### DISCUSSION TITLE
Provide the ability to ignore irrelevant events when converting from Bytes to MIDI

### DISCUSSION THREAD

**Discussion by joaodarocha1** [melanchall/drywetmidi#192] [Ideas]
Method:  BytesToMidiEventConverter.ConvertMultiple(byte[] bytes, int offset, int length)

When converting a stream of device messages to Melanchall.DryWetMidi.Core.MidiEvent we need the ability to select only the relevant events.  A device such as a Yamaha Genos will send too many irrelevant system messages. DryWetMidi will fail to read them and throw too many exceptions. 

It would be great to get only the relevant messages types such as EventStatusBytes.Channel.NoteOn and  EventStatusBytes.Channel.NoteOff.  Once we have the statusByte we could skip the event and move on to the next one.

**Comment by melanchall** [comment]
> A device such as a Yamaha Genos will send too many irrelevant system messages. DryWetMidi will fail to read them and throw too many exceptions

Can you please explain this in more detail? DWM should be able to read all MIDI events allowed by MIDI spec. I suppose your device sends sys ex events, but DWM should successfully read them.

Please show your code, exceptions and all data that can be useful to understand your issue and what's going on.

**Reply by joaodarocha1** [reply]
Upon further debugging, I've noticed that there is a **NotEnoughBytesPolicy** property on the **BytesToMidiEventConverter** class.  Setting this property to NotEnoughBytesPolicy.Ignore has reduced significantly the number of exceptions thrown by the converter. I am still getting some UnexpectedRunningStatusException though.

Please find below a sample of my test code working correctly with the **NotEnoughBytesPolicy** property set to Ignore.


//NoteOn event
byte[] data = Convert.FromBase64String("AZBDL7o7tIIUAQEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==");
var bytesToMidiEventConverter = new BytesToMidiEventConverter();
var events = bytesToMidiEventConverter.ConvertMultiple(data, 1, 3);

//NormalSysEx event
data = Convert.FromBase64String("AfBDfgIxADiTt0hhAQEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==");
bytesToMidiEventConverter = new BytesToMidiEventConverter();
bytesToMidiEventConverter.NotEnoughBytesPolicy = NotEnoughBytesPolicy.Ignore;
events = bytesToMidiEventConverter.ConvertMultiple(data, 1, 6);

Thanks for you help!

**Reply by melanchall** [reply]
Well, regarding NormalSysEx. Bahavior of the DryWetMIDI is absolutely correct. If we look what bytes you have:

![image](https://user-images.githubusercontent.com/26361185/171363308-c8aeb669-9caa-42c5-87d3-10ce7e5479bf.png)

Sys ex event structure is: status byte, then, length of data (VLQ format), then data. The highlighted byte represents the length of an event. It's `67`. But you specify reading from byte 1 using 6 bytes. `6` is not `67` required to read the event, so we have the exception.

If we write this:

```csharp
bytesToMidiEventConverter.ConvertMultiple(data, 1, 69);
```

we'll get sys ex event successfully read.

But yes, if we write this:

```csharp
bytesToMidiEventConverter.ConvertMultiple(data, 1, data.Length - 1);
```

we'll get `UnexpectedRunningStatusException`, but it's the next error. I'm pretty sure the issue is with the data. But I also see that DWM actually is not clear **where** in the bytes array exception occurred so it would be great to improve the library regarding this.

**Comment by joaodarocha1** [comment]
Please find below a sample of MIDI messages from the Yamaha Workstation that throws UnexpectedRunningStatusException.

byte[] data = Convert.FromBase64String("ATEA91Ej7J0vGgEAGgEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==");
var bytesToMidiEventConverter = new BytesToMidiEventConverter();
bytesToMidiEventConverter.NotEnoughBytesPolicy = NotEnoughBytesPolicy.Ignore;
var events = bytesToMidiEventConverter.ConvertMultiple(data, 1, 3);

data = Convert.FromBase64String("ATYe98icBy7eGwEAGwEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==");
bytesToMidiEventConverter = new BytesToMidiEventConverter();
bytesToMidiEventConverter.NotEnoughBytesPolicy = NotEnoughBytesPolicy.Ignore;
var evaents = bytesToMidiEventConverter.Convert(data, 1, 3);

**Reply by melanchall** [reply]
For the first data:

![image](https://user-images.githubusercontent.com/26361185/171429812-5a45b8c4-ca47-4225-bd1d-792cc74c3f14.png)

So status byte is `49` (`0x31`) which means we have running status here (because the number is less than `0x80`). But we have not real status byte before so DWM doesn't know the type of the current event. It sees running status, but it doesn't see status byte specified before. So this piece of data definitelycan't be read. I suppose your device sends data before bytes you've provided me and in that data there is the real status byte. This isolated piece of data is corrupted without context.

**Comment by joaodarocha1** [comment]
I am attaching a log of messages that leads to the UnexpectedRunningStatusException exception. 

[midi sequence.txt](https://github.com/melanchall/drywetmidi/files/8815395/midi.sequence.txt)

**Reply by melanchall** [reply]
I can't say nothing new. You've included a piece of data with wrong sys ex. Please don't ignore NotEnoughBytesException here. Please read again what I wrote you before. Sys ex defines its length as 67, but you allow to read only 6 bytes (why? from where you get these indices and lengths?). And thus all goes wrong.

Last piece of data is corrupted, read again my previous message. I really don't understand how those bytes parts should be read and what you expect.

**Reply by joaodarocha1** [reply]
My app is a Xamarin application. I am using the count parameter of the method [MidiReceiver.onSend](https://developer.android.com/reference/android/media/midi/MidiReceiver#onSend(byte[],%20int,%20int,%20long)).

The documentation says: "count int: the number of bytes of MIDI data in the array to be processed"

**Reply by joaodarocha1** [reply]
Reading a little bit more of the android MidiReceiver documentation:

**"NOTE: the msg array parameter is only valid within the context of this call. The msg bytes should be copied by the receiver rather than retaining a reference to this parameter. Also, modifying the contents of the msg array parameter may result in other receivers in the same application receiving incorrect values in their {link #onSend} method."**

This issue might be related to data corruption indeed. DWM is correct.

**Reply by melanchall** [reply]
And what documentation on the method should say? You specify count manually, passing a number. I ask you about specific numbers. Why you think we need to read 6 bytes for sys ex? Declared length is 67, so why 6?

I also don't understand what these base64 strings means, what relationship between them and so on. I won't be able to help you if you don't provide me detailed info. Can you just give me entire bytes sequence from the device? Parts you've provided me are not related with each other and so we have corruptions. No one library will be able to read data with such corruptions. More than than, in my opinion last part in your _midi sequence.txt_ starts with end of some sys ex event. Well, if these bytes are the end of an event, where its start? We need parts chronologically sorted.

**Reply by joaodarocha1** [reply]
base64 strings are just an easy way to share data here on GitHub messages. 

I've included a link on my previous message to the documentation of the method [MidiReceiver.onSend](https://developer.android.com/reference/android/media/midi/MidiReceiver#onSend(byte%5B%5D,%20int,%20int,%20long)). https://developer.android.com/reference/android/media/midi/MidiReceiver#onSend(byte[],%20int,%20int,%20long)

The documentation of the parameter **count int:** says: "the number of bytes of MIDI data in the array to be processed". I get 6 as the number of bytes to be processed, not 67.

**Reply by melanchall** [reply]
Can you please dump **all the data** as decimal or hex numbers to a text file? Not by base64 parts, but by list of numbers. Content of the file will be something like that:

```
64 23 00 45 ...
```

And then we will be able to talk about the data in more detail. It seems partition by base64 slices causes problems.

**Reply by melanchall** [reply]
@joaodarocha1 Any news?

**Reply by joaodarocha1** [reply]
I will provide the log soon.

**Reply by joaodarocha1** [reply]
Here is the dump of the data that leads to the exception
[dump.txt](https://github.com/melanchall/drywetmidi/files/8856897/dump.txt)
.

**Reply by melanchall** [reply]
Well, this dump is almost the same you've sent me before as base64. My asking was about full dump with all bytes as they coming from the device. What are these blocks of 1024 bytes coming from? Can you please send me the data BEFORE this chunking on blocks?

As for dump.txt, of course, the problem the same as with base64 – the data (bytes along with offsets and lengths) is broken, both on sys ex message and the last one. My assumption is the chunking made wrong, causing the data corruption.

**Reply by joaodarocha1** [reply]
I am running the code on an android device. The blocks of 1024 bytes are created by android MIDI API. 

https://developer.android.com/reference/android/media/midi/package-summary

I am using the following device to convert MIDI to USB

http://www.funs-tech.com/h-pd-17.html?fromColId=104

**Reply by melanchall** [reply]
@joaodarocha1 Do you mean `MidiReceiver` class receives the MIDI data by blocks of 1024 bytes? If so, are you sure you gave me all captured blocks as they came from the device (so no blocks missing)?

**Reply by joaodarocha1** [reply]
This is the log that I got from another tool. Is it of any help to you? 
[midi raw log.txt](https://github.com/melanchall/drywetmidi/files/8952661/midi.raw.log.txt)

**Comment by melanchall** [comment]
@joaodarocha1 Thanks for the last log. Well, it's my fault, I forgot a small thing – `BytesToMidiEventConverter` performs conversions according to Standard MIDI Files (SMF) specification which states:

> <sysex event> is used to specify a MIDI system exclusive message, either as one unit or in
> packets, or as an "escape" to specify any arbitrary bytes to be transmitted. A normal complete
> system exclusive message is stored in a MIDI File in this way:
> 
> F0 `<length>` `<bytes to be transmitted after F0>`

And that's why DryWetMIDI tries to read sys ex event length although there is no length there since the data got from a device, not from MIDI file. Right now there is no way to convert sys ex bytes without length with `BytesToMidiEventConverter`. Probably I need to improve converters providing a kind of `UseSmfFormat` property which is `true` by default, but can be set to `false`.

But as for last data

> 31 22 F7: data()

It's strange anyway. It looks like it's the end of previous sys ex

> F0 43 7E 02 31 22: SysEx()

But why last bytes are sent seperately along with `0xF7`... I don't know about such things.

**Reply by melanchall** [reply]
@joaodarocha1 Well, I've published prerelease version of the library – [6.1.2-prerelease1](https://www.nuget.org/packages/Melanchall.DryWetMidi/6.1.2-prerelease1). Now you can set `BytesFormat` property on converters to `BytesFormat.Device` which switches how sys ex (and some other events) are read/written.

**Reply by joaodarocha1** [reply]
Awesome! I will test it.

**Comment by melanchall** [comment]
:rocket: **6.1.2 version is released now!**

Prerelease NuGet packages will be unlisted soon, so please update the package references to the new version.

Thanks for using the library!

**Comment by joaodarocha1** [comment]
Sorry for the late reply. I was not working on this project for a while.

It is working now. I've added a video of the app working connect to the Yamaha keyboard.

https://github.com/melanchall/drywetmidi/assets/48954066/1ef6dceb-bbdc-4da7-bb63-dc5f56d971ee
